### DISCUSSION TITLE
How to update and go to setup mode through home assistant?

### DISCUSSION THREAD

**Discussion by alialki** [RavenSystem/esp-homekit-devices#1931] [HAA Home App]
I am both new to HAA and Home assistant could anyone tell me how I can update firmware through home assistant homekit controller, as the devices are no longer exposed to HAA manager app through the bridge i created there is an entity for update defined as switch but when pressed nothing happens at all. 

I would really appreciate it if some one can help me.

**Comment by ondratan** [comment]
Have you tried going to device setup and adding the device again?

**Comment by alialki** [comment]
To be honest i wanted to avoid that route for the time being as it will be really time consuming

**Comment by euhkah** [comment]
Hello, I'm looking for the same solution.

Automations in homekit are for the moment without interest, without conditions.

My devices use HAA firmware, but they are configured with Home Assistant.
I bought the HAA Manager application, but I cannot activate Setup Mode and update his.

It is necessary to update HAA Manager to make this possible, it is very important ! @RavenSystem

**Comment by euhkah** [comment]
[In addition to the problem of unable to update, there is also the inability to modify the wifi configuration, the MEPLHAA configuration, etc.]

**Comment by alialki** [comment]
I bought the HAA Manager app as well, I hope @RavenSystem updates the HAA manager to make this possible too. At the moment the best solution I found is to toggle the the device 8 times as fast as possible to take the devices in to setup mode then the app setup radar feature pops it up and you can open it in safari and change settings and update firmware from there. But still this is not an efficient solution specially for light switches, it will lower the light bulbs life span

**Comment by alialki** [comment]
@euhkah have you managed to find a better solution than mine ? Because with my workaround i have blown two light bulbs since 10 days ago

**Comment by euhkah** [comment]
I couldn't find anything better. I applied your solution to switch my lights to Homekit... temporarily. I'm thinking of going back to Tasmota, even if it means plugging in the wires. Especially since there is no form of support or help here.

**Comment by freedreamer82** [comment]
it should be a very nice feature.
I also bought the app and every time I have to remove from HA the device , pair it back to my phone and update.... :( very long process ....
Please make it upgradable from  HA ( even if I I suspect is a problem HA side).

Marco

**Comment by euhkah** [comment]
Hi Marco, as alialki said, you turn the device on/off several times. It will go into setup mode. By clicking on the app's radar, you can see your device, open its configuration and update it, without deleting it... But this procedure is also very long !
This firmware is not suitable for home assistant. It is better to put Tasmota if you don't use Homekit natively.

**Comment by freedreamer82** [comment]
@euhkah no way , it's better haa than tasmota.
actually on home assistant there is a buttton to go on update setup mode the device...but i cannot get it working...
https://github.com/home-assistant/core/pull/59750

**Comment by freedreamer82** [comment]
@euhkah I fixed the HA stuff.. once I'll be ready I'll make a PR on hass to have the chance to have button for UPDATE/SETUP in hass UI.
Regards 

Marco

**Comment by euhkah** [comment]
That would be great! A beautiful Christmas present!

Because I also have the update button that doesn't do anything...

**Comment by freedreamer82** [comment]
@euhkah https://github.com/home-assistant/core/pull/84502
here you can find the mod if you wanna try.
cheers

**Comment by RavenSystem** [comment]
HAA uses custom HomeKit services and characteristics to manage setup features. For security reasons, HAA running in normal mode only supports HomeKit for all interactions. There are not hidden protocols. HAA Home Manager App sends commands to HAA devices through HomeKit.

When a HAA device is added to HomeAssistant, this last must publish and passthrough custom HAA services and characteristics to allow HAA Home Manager App to work properly.

As far I know, HomeAssistant HomeKit support is very limited, and it doesn't support passthrough of custom services. A way to solve it is declaring standard switches into HAA script with system actions to setup, update, reboot and wifi reconnection.

**Reply by freedreamer82** [reply]
Yep that’s exactly what the pr i Made in
Ha does. Home Assistant support for homekit is very limited , btw having setup Button is enough to work with haa devices.
(Update etc..)
Thanks

Marco

**Reply by RavenSystem** [reply]
Great! A passthrough is the best solution because I am constantly include changes in HAA setup service. In this way, HAA Home Manager App will can send right commands directly and HA will pass it as is to HAA device.

**Reply by freedreamer82** [reply]
I tried to ask a passthrough to use with HAA app but is not possible in home assistant architecture... :(
In the future , if you can , before changing the way to enter in setup mode consider that could break compatibility :)

**Reply by RavenSystem** [reply]
Passthrough should be the way to implement. Sorry, but I can not keep the same way to enter setup mode because command string will be changed with some versions to avoid piracy of HAA Manager App. I can guarantee that HAA Setup Service is and will be HAP (HomeKit Accessory Protocol) compatible, but nothing more.

**Reply by freedreamer82** [reply]
With all respect , I see your point as HAA Manager developer...but i guess It's not the way to go. Since you released haa as open source you should guarantee in some way any kind of integration in other system.Home assistant is just an example ... but if anyone would like to integrate your work somewhere else with this restrictions you have been setting is very hard. 
Maybe you should find another way to avoid piracy in your app...
 
just my 2 cent.

M.

**Reply by RavenSystem** [reply]
To clarify: Source code is available, but HAA is not open source; it is a firmware with native HomeKit support that I develop for my needs, and I share with others that want to use it. It is designed as a fully customizable system following Apple HAP specs. HAP covers all my needs (and 99%, even more, of the needs of others), without having a central server (it is how distributed systems work for high availability). If some people want to use HAA with HomeAssistant, this last should follow Apple HAP specs, like HAA. A HAP string characteristic can send or receive any string; if HomeAssistant is not able to manage it, its developers should implement HAP in a right way. HAA follows HAP specs, HAA Home Manager App follows HomeKit SDK specs, and both works perfectly together. Probably, best solution for HomeAssistant is to flash ESPHome (officially supported by HomeAssistant) instead HAA. Other solution is to re-pair HAA devices directly with iOS using Extra Pairing feature, and adding them in a new house into Apple Home App to keep them separated from HomeAssistant HAA devices, letting use HAA Manager App from this last house.

**Reply by freedreamer82** [reply]
I see... the solution of extra pairing could be an idea , I have to understand if i can re-import them from ios to HA. 
BTW I'm trying to bypass home assistant writing directly to the characteristic using the low level library. If I have any updates I'll write it back.

Marco

**Reply by RavenSystem** [reply]
Yes, you can pair HAA with iOS and Home Assistant simultaneously using Extra Pairing feature.

**Reply by freedreamer82** [reply]
mmmh if it works with multiple controller could be really the solution.I'll try and let u know.

**Reply by freedreamer82** [reply]
it works. Good point...for anyone who is interested to add in home Assistant I guess is the solution for now. The only one concern is having two controllers u have more traffic in the network , since as far as i know Homekit is in polling.

Thanks.

**Reply by RavenSystem** [reply]
You have more traffic, but HomeKit doesn't use a polling system (it is a bad practice), and traffic is very, very low. Devices notify events to controllers (when a specific characteristic changes, with ev flag), instead controllers poll devices.

**Comment by freedreamer82** [comment]
just to let know everyone is interested..
https://github.com/home-assistant/core/pull/84519

there is an official PR  from HA team started from mine ...but in the last version of HAA there are still changes in how to enter in setup mode so they have been thinking how to proceed for the old device compatibility.


cheers

M.

**Reply by RavenSystem** [reply]
You are right. Since v10.0.0, HAA setup service changed from 2 characteristics to only one managed by commands, adding more features and saving DRAM. Probably, best way is to use HAA version to trigger old system or pass HAA Manager APP command directly to HAA device.

**Comment by freedreamer82** [comment]
@alialki  For those who are using home assistant I created this project. My suggestion as written is to use the extra pairing in order to use APP Manager ... however for those who are not using an apple TV or home hub maybe be useful.
Bye

https://github.com/freedreamer82/haa_manager_cli

**Reply by RavenSystem** [reply]
Great!

FYI: HomeKit works without AppleTV or HomeHUB in LAN. It is a distributed system.

**Reply by freedreamer82** [reply]
yes , you are right.An apple TV is for remote access.

**Comment by alialki** [comment]
@freedreamer82 thank you very much

**Reply by freedreamer82** [reply]
you are welcome. Try it and let me know.
